### ISSUE TITLE
java.lang.UnsatisfiedLinkError: 'org.bridj.Pointer com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.getVideoDevices()

### ISSUE THREAD

**Issue by emmanuelrosa** [sparrowwallet/sparrow#294]
When I attempt to scan a QR code I get the following stacktrace:

```
com.github.sarxos.webcam.WebcamException: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService.getWebcams(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.Webcam.getWebcams(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamService$1.call(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamService$1.call(Unknown Source)
	at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at javafx.graphics/javafx.concurrent.Service.lambda$executeTask$6(Service.java:725)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at javafx.graphics/javafx.concurrent.Service.lambda$executeTask$7(Service.java:724)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	... 12 common frames omitted
Caused by: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamTask.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$GetDevicesTask.getDevices(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver.getDevices(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamScanDriver.getDevices(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 common frames omitted
Caused by: java.lang.UnsatisfiedLinkError: 'org.bridj.Pointer com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.getVideoDevices()'
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.getVideoDevices(Native Method)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$GetDevicesTask.handle(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.run(Unknown Source)
	... 3 common frames omitted
2021-10-30 10:55:21,976 ERROR Error scanning QR
com.github.sarxos.webcam.WebcamException: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService.getWebcams(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.Webcam.getWebcams(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamService$1.call(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamService$1.call(Unknown Source)
	at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at javafx.graphics/javafx.concurrent.Service.lambda$executeTask$6(Service.java:725)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at javafx.graphics/javafx.concurrent.Service.lambda$executeTask$7(Service.java:724)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	... 12 common frames omitted
Caused by: com.github.sarxos.webcam.WebcamException: Cannot execute task
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamTask.process(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$GetDevicesTask.getDevices(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver.getDevices(Unknown Source)
	at com.sparrowwallet.sparrow/com.sparrowwallet.sparrow.control.WebcamScanDriver.getDevices(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 common frames omitted
Caused by: java.lang.UnsatisfiedLinkError: 'org.bridj.Pointer com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.getVideoDevices()'
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.getVideoDevices(Native Method)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$GetDevicesTask.handle(Unknown Source)
	at com.github.sarxos.webcam.capture@0.3.13-SNAPSHOT/com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.run(Unknown Source)
	... 3 common frames omitted
```

### System info:

```
uname -a 
Linux abc123 5.10.76 #1-NixOS SMP Wed Oct 27 07:56:57 UTC 2021 x86_64 GNU/Linux

nixos-version
21.11.20211029.8d380db (Porcupine)
```

**Comment by emmanuelrosa** [comment]
Theses are the libraries linked to OpenIMAJGrabber. I'm curious whether my system has some libraries which may be too new and causing an ABI incompatability.

```
ldd /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
    linux-vdso.so.1 (0x00007ffd6e18e000)
    libv4l2.so.0 => /nix/store/...-v4l-utils-1.20.0/lib/libv4l2.so.0 (0x00007fc71be57000)
    libstdc++.so.6 => /nix/store/...-gcc-10.3.0-lib/lib/libstdc++.so.6 (0x00007fc71bc82000)
    libgcc_s.so.1 => /nix/store/...-gcc-10.3.0-lib/lib/libgcc_s.so.1 (0x00007fc71bc68000)
    libc.so.6 => /nix/store/...-glibc-2.33-50/lib/libc.so.6 (0x00007fc71baa3000)
    libpthread.so.0 => /nix/store/...-glibc-2.33-50/lib/libpthread.so.0 (0x00007fc71ba83000)
    libdl.so.2 => /nix/store/...-glibc-2.33-50/lib/libdl.so.2 (0x00007fc71ba7c000)
    libv4lconvert.so.0 => /nix/store/...-v4l-utils-1.20.0/lib/libv4lconvert.so.0 (0x00007fc71ba00000)
    libm.so.6 => /nix/store/...-glibc-2.33-50/lib/libm.so.6 (0x00007fc71b8bf000) 
    /nix/store/...-glibc-2.33-50/lib64/ld-linux-x86-64.so.2 (0x00007fc71c070000)
    librt.so.1 => /nix/store/...-glibc-2.33-50/lib/librt.so.1 (0x00007fc71b8b4000)
    libjpeg.so.62 => /nix/store/...-libjpeg-turbo-2.1.0/lib/libjpeg.so.62 (0x00007fc71b805000)
```

- I tried enabling Bridj verbose logging, to get more details, but it didn't work. 
- I've read through the OpenIMAJGrabber source code and could not find anything that would suggest that the Java binding is incompatible with the native implementation.
- I did confirm that the OpenIMAJGrabber.so library is loading successfully; I checked the `/proc/n/maps` file of the JVM process.

```
7f53963da000-7f53963df000 r-xp 00000000 00:1b 2756343                    /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
7f53963df000-7f53965de000 ---p 00005000 00:1b 2756343                    /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
7f53965de000-7f53965df000 r--p 00004000 00:1b 2756343                    /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
7f53965df000-7f53965e0000 rw-p 00005000 00:1b 2756343                    /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
7f53965e0000-7f53965e1000 rw-p 0000d000 00:1b 2756343                    /tmp/BridJExtractedLibraries4655314859887623497/OpenIMAJGrabber.so
7f53965e1000-7f53965fc000 r-xp 00000000 00:1b 2756342                    /tmp/BridJExtractedLibraries4655314859887623497/libbridj.so
7f53965fc000-7f53967fb000 ---p 0001b000 00:1b 2756342                    /tmp/BridJExtractedLibraries4655314859887623497/libbridj.so
7f53967fb000-7f53967fc000 r--p 0001a000 00:1b 2756342                    /tmp/BridJExtractedLibraries4655314859887623497/libbridj.so
7f53967fc000-7f53967fd000 rw-p 0001b000 00:1b 2756342                    /tmp/BridJExtractedLibraries4655314859887623497/libbridj.so
```

**Comment by craigraw** [comment]
Thanks for this unusually thorough report! I've haven't yet had a reported webcam issue on Linux.

It may be useful to see if this error can be reproduced outside of Sparrow. The [distributable](https://github.com/sarxos/webcam-capture/releases/download/webcam-capture-parent-0.3.10/webcam-capture-0.3.10-dist.zip) for the webcam project at http://webcam-capture.sarxos.pl/ contains a number of example programs. 

Sparrow uses an updated version of some of these libraries, mainly to handle issues on MacOS. However they are relatively similar. I got it running with the following (JDK 16 installed in my case):

```shell
cd webcam-capture-0.3.10-dist/examples
javac -cp ../webcam-capture-0.3.10.jar:. WebcamViewerExample.java
java -cp ../webcam-capture-0.3.10.jar:../libs/slf4j-api-1.7.2.jar:../libs/bridj-0.6.2.jar:. WebcamViewerExample
```

**Comment by emmanuelrosa** [comment]
I was able to reproduce the error using the WebcamViewerExample in the webcam project.

Thank you.

**Comment by emmanuelrosa** [comment]
I fixed the problem.

NixOS doesn't follow the Linux FHS, so dynamically-linked executables and libraries from other distros have to "relinked" so that `ld` can load them. We have a tool for that, and `ldd` didn't report any issues with OpenIMAJGrabber.so after fixing the links, yet I kept getting this error.

I went on a hunch and compiled OpenIMAJGrabber.cpp on NixOS and used the compiled OpenIMAJGrabber.so to replace the one from upstream. And, it worked! I was able to scan a QR code :)
